processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -267,55 +291,160 @@ static double estimatePredicateSelectivity(
   }
 
   static double estimatePredicateSelectivity(
+      BitmapIndex bitmapIndex,
       ColumnSelector columnSelector,
       String dimension,
-      Iterable<ImmutableBitmap> bitmaps,
+      Iterable<Integer> bitmapIndexes,
       long totalNumRows
   )
   {
     final ColumnCapabilities columnCapabilities = columnSelector.getColumn(dimension).getCapabilities();
-    return estimatePredicateSelectivity(
-        bitmaps,
+    return estimateSelectivityOfBitmapList(
+        bitmapIndex,
+        bitmapIndexes,
         totalNumRows,
+        // assume multi-value column if columnCapabilities is null
         columnCapabilities == null || columnCapabilities.hasMultipleValues()
     );
   }
 
-  static double estimatePredicateSelectivity(
+  private static double estimateSelectivityOfBitmapList(
+      BitmapIndex bitmapIndex,
+      Iterable<Integer> bitmapIndexeIterable,
+      long totalNumRows,
+      boolean isMultiValueDimension
+  )
+  {
+    final List<Integer> bitmapIndexes = ImmutableList.copyOf(bitmapIndexeIterable);
+    long numMatchedRows = 0;
+    for (Integer index : bitmapIndexes) {
+      final ImmutableBitmap bitmap = bitmapIndex.getBitmap(index);
+      numMatchedRows += bitmap.size();
+    }
+
+    if (isMultiValueDimension) {
+      final double estimated = numMatchedRows * computeNonOverlapRatioFromRandomBitmapSamples(
+          bitmapIndex,
+          bitmapIndexes
+      ) / totalNumRows;
+      return Math.min(1., estimated);
+    } else {
+      return (double) numMatchedRows / totalNumRows;
+    }
+  }
+
+  static double estimateSelectivityOfBitmapTree(
       Iterable<ImmutableBitmap> bitmaps,
       long totalNumRows,
       boolean isMultiValueDimension
   )
   {
     long numMatchedRows = 0;
-    final List<ImmutableBitmap> bitmapList = Lists.newArrayList(bitmaps);
-    for (ImmutableBitmap bitmap : bitmapList) {
+    for (ImmutableBitmap bitmap : bitmaps) {
       numMatchedRows += bitmap.size();
     }
 
-    // assume multi-value column if columnCapabilities is null
     if (isMultiValueDimension) {
-      final double estimated = numMatchedRows * Filters.computeNonOverlapRatioFromBitmapSamples(bitmapList)
-                               / totalNumRows;
+      final double estimated = numMatchedRows * computeNonOverlapRatioFromFirstNBitmapSamples(
```

Comment:
It seems to me that the math model of this expression and `computeNonOverlapRatioFromRandomBitmapSamples()`  is wrong. Imagine there are 10 bitmaps each of 0.1 selectivity, and they all overlap 100%. `computeNonOverlapRatioFromRandomBitmapSamples()` will return 0.5, estimated will be 10 * 0.1 * 0.5 = 0.5, while it should be 0.1.